### DISCUSSION TITLE
changing modes without clearing grid's script?

### DISCUSSION THREAD

**Discussion by dndrks** [monome/iii#2] [Ideas]
holding key `0,0` while booting has three uses:

- if there's no script uploaded / a blank script (plasma), grid will cycle to serial mode
- if grid is in serial mode, grid will cycle to plasma (no script / blank)
- if there is a script uploaded, grid will clear it and cycle to plasma
  - from plasma, one more hold+hotplug will cycle grid to serial mode

this accommodates cases where one wants to get to serial mode _or_ where one wants to clear the currently-running script, but i can imagine a situation where one might want to simply change into serial mode from a grid-script (eg. i started the show with `es.lua` and now this part of the set uses `awake` on norns, so i need grid to be in serial mode to control it) and then *change back* to the previously-running script (eg. this part of the set needs `es.lua` again).

currently, cycling from a grid-script to serial necessitates _erasing_ the script from memory -- i'm curious if it's technically possible to toggle between a script and serial mode without losing the script?

**Comment by tlubke** [comment]
Was thinking about this too as I imagined how I would be using it.

Maybe there could be a dedicated _erase_ key, say `15,15` that is held in combination with `0,0` to erase that script. Would be cool each key aside from the _erase_ key could hold a script, but maybe flash space is too limited? I could see myself writing a bunch of different MIDI mappings as tiny lua scripts. Being able to store them on the device and boot into them with a shortcut would be nice.

**Comment by tehn** [comment]
just to change directions a bit--- i'd like to maintain consistency with the (upcoming) arc which has one pushbutton, so i think this pattern must hold.

_however_ i am now thinking it'd make sense detect in the serial stream if serialosc/etc are querying the serial setup bytes (ie, grid size... which serialosc, norns, and ansible all do)... upon seeing that byte sequence it can magically toggle modes (and soft reset)

this actually solves another problem as well which is serialosc trying to grab iii devices (which could be solved other ways easily as well)

**Reply by dndrks** [reply]
the one-button constraint makes a lot of sense -- is there any possibility of nestling the options inside of a "launch" mode? eg. if the button is held while powering, then we're in a 'menu' where:
- the button is pressed once: launch serial mode
- the button is pressed twice: launch stored iii script
- the button is held: clear iii script and cycle to blank plasma mode

is ^ that at all aligned?

---
also just to confirm my reading of this:

> however i am now thinking it'd make sense detect in the serial stream if serialosc/etc are querying the serial setup bytes (ie, grid size... which serialosc, norns, and ansible all do)... upon seeing that byte sequence it can magically toggle modes (and soft reset)

does this mean that if an otherwise-running-iii-script grid is connected to norns, it will automatically toggle modes (and soft reset = erase the iii script)? i guess i'm hoping for a less 'destructive' mechanism, where flipping between serial mode and iii mode _doesn't_ nuke the stored script.

**Comment by tehn** [comment]
indeed i realized that connecting to norns/ansible fail the approach i suggested due to there being no way to "disable" the serialosc mode.

i think the menu mode proposition works. i conceived of a couple other options i'm going to try as well.

**Comment by tehn** [comment]
new approach:

- hold key to switch modes
- when switching into iii mode, if you hold for 2sec it will erase the script
